ethcore/src/engines/clique/block_state.rs

```diff
@@ -0,0 +1,206 @@
+	pub fn apply(&mut self, header: &Header, is_checkpoint: bool) -> Result<Address, Error> {
+		let creator = self.verify(header)?;
+
+		// rotate recent signers.
+		self.recent_signers.push_front(creator.clone());
+		if self.recent_signers.len() >= ( self.signers.len() / 2 ) + 1 {
+			self.recent_signers.pop_back();
+		}
+
+		if is_checkpoint {
+			// checkpoint block should not affect previous tallying, so we check that.
+			let signers = extract_signers(header)?;
+			if self.signers != signers {
+				return Err(From::from("checkpoint block signers is different than expected"));
+			};
+
+			self.votes.clear();
+			self.votes_history.clear();
+
+			// maybe release some memory.
+			self.votes.shrink_to_fit();
+			self.votes_history.shrink_to_fit();
+
+			return Ok(creator);
+		}
+
+		let beneficiary = *header.author();
+
+		// No vote, ignore.
+		if beneficiary == NULL_AUTHOR {
+			return Ok(creator);
+		}
+
+		let nonce = header.decode_seal::<Vec<&[u8]>>().unwrap()[1];
+
+		let vote_type: VoteType;
+
+		if NONCE_AUTH_VOTE == nonce {
+			vote_type = VoteType::Add;
+		} else if NONCE_DROP_VOTE == nonce {
+			vote_type = VoteType::Remove;
+		} else {
+			return Err(From::from("beneficiary specified but nonce was not AUTH or DROP"));
+		};
+
+		// Record this vote, also since we are using an hashmap, it will override previous vote.
+		self.votes.insert((creator.clone(), beneficiary), vote_type);
+		self.votes_history.push((header.number(), creator.clone(), vote_type, beneficiary));
+
+		// Tally up current target votes.
+		let threshold = self.signers.len() / 2;
+		let vote = self.votes.iter().filter(|(key, value)| {
+			(**key).1 == beneficiary && **value == vote_type
+		}).count();
+
+		if vote > threshold {
+			match vote_type {
+				VoteType::Add => {
+					self.signers.push(beneficiary);
+				},
+				VoteType::Remove => {
+					let pos = self.signers.binary_search(&beneficiary);
+					if pos.is_ok() {
+						self.signers.remove(pos.unwrap());
+					}
+				}
+			}
+
+			self.signers.sort();
+
+			// Remove all votes about or made by this beneficiary
+			{
+				let votes_copy = self.votes.clone();
+				let items: Vec<_> = votes_copy.iter().filter(|(key, _value)| {
+					(**key).0 == beneficiary || (**key).1 == beneficiary
+				}).collect();
+
+				for (key, _) in items {
+					self.votes.remove(&key);
+				}
+			}
+		}
+
+		// No cascading votes.
+		Ok(creator)
+	}
+
+	pub fn calc_next_timestamp(&mut self, header: &Header, period: u64) {
+		let base_time = UNIX_EPOCH + Duration::from_secs(header.timestamp());
+
+		self.next_timestamp_inturn = Some(base_time + Duration::from_secs(period));
+
+		let delay = Duration::from_millis(
+			rand::thread_rng().gen_range(0u64, (self.signers.len() as u64 / 2 + 1) * SIGNING_DELAY_NOTURN_MS));
+		self.next_timestamp_noturn = Some(base_time + Duration::from_secs(period) + delay);
+	}
+
+	pub fn inturn(&self, current_block_number: u64, author: &Address) -> bool {
```

Comment:
nit
```suggestion
	pub fn in_turn(&self, current_block_number: u64, author: &Address) -> bool {
```

Reply:
this was named after geth impl's function. but I changed to is_inturn now.